src/finn/custom_op/fpgadataflow/iodma.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+from onnx import TensorProto, helper
+from finn.core.datatype import DataType
+from finn.custom_op.fpgadataflow import HLSCustomOp
+
+
+class IODMA(HLSCustomOp):
+    """Class that corresponds to finn-hlslib DMA function(s)."""
+
+    def __init__(self, onnx_node):
+        super().__init__(onnx_node)
+
+    def get_nodeattr_types(self):
+        my_attrs = {
+            "NumChannels": ("i", True, 0),
+            # FINN input datatype
+            "dataType": ("s", True, ""),
+            # DMA-specific parameters
+            "intfWidth": ("i", False, 32),
+            "burstMode": ("s", False, "increment"),
+            "direction": ("s", False, "in"),
+            # shape describing input vecs per execution
+            "numInputVectors": ("ints", False, [1]),
+        }
+        my_attrs.update(super().get_nodeattr_types())
+        return my_attrs
+
+    def get_normal_input_shape(self):
+        vecs = list(self.get_nodeattr("numInputVectors"))
+        num_ch = self.get_nodeattr("NumChannels")
+        ishape = tuple(vecs + [num_ch])
+        return ishape
+
+    def get_normal_output_shape(self):
+        return self.get_normal_input_shape()
+
+    def get_folded_input_shape(self):
+        shape = list(self.get_normal_input_shape())
+        itype_bits = self.get_input_datatype().bitwidth()
+        intfw = self.get_nodeattr("intfWidth")
+        elems_per_word = intfw / itype_bits
+        fold_depth = round(shape[-1] / elems_per_word)
+        shape[-1] = fold_depth
+        shape.append(elems_per_word)
+        return tuple(shape)
+
+    def get_folded_output_shape(self):
+        return self.get_folded_input_shape()
```

Comment:
shouldn't the input and output folded shapes be different for the stream/mem DMA engines? or is it always guaranteed that stream width = intf width?

Reply:
In fact folded shape can't reliably be defined for the axi-mm interface. Modified the code to define the folded shape on the stream side properly, since that is the side connected to any other FINN logic.